### PR TITLE
Adds full support for Skaffold profiles selection for run/dev

### LINKED ISSUE #51: Support Skaffold profile selection for run configurations
Skaffold supports multiple `profiles` in a single configuration file, for example, one for docker build, and one for Cloud Build, see https://github.com/GoogleContainerTools/skaffold/blob/master/examples/getting-started/skaffold.yaml with two additional profiles (`gcb`, `test`).

We can parse the configuration file and find all profiles. If more than one profile (default) exists, we then provide a drop down with the list of profiles and allow a user to select which profile needs to be run. Chosen profile is then passed on to `skaffold` with a `-p --profile` flag.

### PR DESCRIPTION
**Pull request by ivanporty**
Fixes #51.

Introduces profile selection combobox, which by default uses `default` profile, i.e. default Skaffold configuration settings. When no additional profiles are defined, combo box is disabled.

skaffold-default-profile

When more profiles are available, combo box becomes enabled and allows choosing which profile to select and develop/deploy with:

skaffold-profile-selection

Resulting CLI command with selected profile:

skaffold-profile-cli

Selected profile is persisted in settings and can be different for different profiles.

### CODE REVIEW

**Comment by etanshaul** [comment]
+ @iantalarico (I think I remember you asking what we plan to do with profiles once ..)

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold
+
+import com.intellij.openapi.vfs.VirtualFile
+import org.yaml.snakeyaml.Yaml
+import java.io.ByteArrayInputStream
+
+class SkaffoldYamlConfiguration(skaffoldYamlFile: VirtualFile) {
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt]: javadoc

Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt]: *kotlindoc

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    private val skaffoldYamlMap = mutableMapOf<Any, Any>()
+
+    init {
+        val yamlLoader = Yaml()
+        skaffoldYamlMap.putAll(
+            yamlLoader.load(ByteArrayInputStream(skaffoldYamlFile.contentsToByteArray())))
+    }
+
+    val profiles: Map<String, Any>
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt]: some documentation here would be nice too

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold
+
+import com.intellij.openapi.vfs.VirtualFile
+import org.yaml.snakeyaml.Yaml
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt]: just wondering, where did this dependency come from? is it part of the yaml plugin?

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    fun setSelectedProfile(profile: String) {
+        for (i in 0 until model.size) {
+            if (model.getElementAt(i).equals(profile)) {
+                model.selectedItem = profile
+                break
+            }
+        }
+    }
+
+    fun skaffoldFileUpdated(skaffoldFile: VirtualFile?) {
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt]: doc would be helpful

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt (hunk 2 above)
Comment (etanshaul) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt]: perhaps with example of what a yaml with a profile could look like

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import com.google.common.annotations.VisibleForTesting
+import com.google.container.tools.skaffold.SkaffoldYamlConfiguration
+import com.google.container.tools.skaffold.message
+import com.intellij.openapi.diagnostic.Logger
+import com.intellij.openapi.vfs.VirtualFile
+import javax.swing.DefaultComboBoxModel
+import javax.swing.JComboBox
+
+/**
+ * Combobox for selecting Skaffold profiles. Default profile is assumed to be always present for a
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt]: by "assumed to be always present" do you mean that it is implicitly there always (even when no profile is defined in the yaml)?

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    init {
+        setModel(model)
+        isEnabled = false
+    }
+
+    fun getSelectedProfile(): String? {
+        return if (selectedIndex <= 0) null else model.getElementAt(selectedIndex)
+    }
+
+    /** Selects profile if it exists in the profile list. Non existing profile is ignored. */
```
Comment (etanshaul) [inline skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt]: I wonder why github is being so weird with the colorizing on kotlin files...

**Comment by iantalarico** [comment]
@etanshaul Thanks for the info!

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt (hunk 3 above)
Comment (ivanporty) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfiguration.kt]: Correct, part of yaml/IDE.

File: skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt (hunk 5 above)
Comment (ivanporty) [reply skaffold/src/main/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBox.kt]: Right, if Skaffold file is correct, it is itself presents a "default profile". Perhaps I can slightly rephrase to make it more obvious.

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+        profilesComboBox.skaffoldFileUpdated(skaffoldYamlFile)
+
+        assertThat(profilesComboBox.model.size).isEqualTo(1)
+        assertThat(profilesComboBox.model.getElementAt(0)).isEqualTo("default")
+        assertThat(profilesComboBox.isEnabled).isFalse()
+    }
+
+    @Test
+    @UiTest
+    fun `skaffold yaml without profiles results in null profile selection`() {
```
Comment (etanshaul) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: this might just be a wording thing, but reading through this had me confused:

the previous test case says that a skaffold yaml without profiles results in a "...default profile", while this one says it results in a "...null profile..".

Maybe reword if its just a wording thing?

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+        val profilesComboBox = SkaffoldProfilesComboBox()
+
+        profilesComboBox.skaffoldFileUpdated(skaffoldYamlFile)
+
+        assertThat(profilesComboBox.getSelectedProfile()).isNull()
+    }
+
+    @Test
+    @UiTest
+    fun `skaffold yaml with profiles results in combobox with valid profile selection`() {
```
Comment (etanshaul) [inline skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: I guess implicit in here is that when there are multiple profiles, then "default" will be selected by default?

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfigurationTest.kt [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        )
+
+        val skaffoldYamlConfiguration = SkaffoldYamlConfiguration(skaffoldYamlFile)
+
+        assertThat(skaffoldYamlConfiguration.profiles).isNotEmpty()
+        assertThat(skaffoldYamlConfiguration.profiles.keys).isEqualTo(setOf("localImage", "gcb"))
+    }
+
+    @Test
+    fun `invalid yaml file results in an exception`() {
```
[the remark this answers is not in the document]
Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfigurationTest.kt]: Probaby a good idea to add another test case for this.

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldSingleRunSettingsEditorTest.kt [hunk 10]
```diff
@@ -53,7 +54,8 @@ class SkaffoldSingleRunSettingsEditorTest {
     @Before
     fun setUp() {
         EdtTestUtil.runInEdtAndWait(ThrowableRunnable {
-            singleRunSettingsEditor = SkaffoldSingleRunSettingsEditor()
+            singleRunSettingsEditor =
+                SkaffoldSingleRunSettingsEditor()
```
[the remark this answers is not in the document]
Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/SkaffoldSingleRunSettingsEditorTest.kt]: thanks, must be something auto tool glitch.

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt (hunk 8 above)
Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: Hmm.. not really a purpose of this test case to test what is the resulting selection, but rather a list of profiles to choose from is correct. How about
`skaffold yaml with profiles produces valid list of profiles to select from`

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt (hunk 7 above)
Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: Updated and also added a comment on `getSelectedProfile` for clarity.

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfigurationTest.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfigurationTest.kt]: done, it will result in an exception too.

From user perspective, we shouldn't allow this file to bubble up the exception - caller must ensure he handles possible parsing errors when using this class. Pretty much same model as with most of the I/O / parsing classes.

[the remark this answers is not in the document]
Comment (etanshaul) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/SkaffoldYamlConfigurationTest.kt]: great. Thats what I was getting at, just making sure that the callers handle this properly so that the UI works as expected

File: skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt (hunk 8 above)
Comment (etanshaul) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: wording sgtm. Is there another test case to test which selection is defaulted? (sorry if you have one already and I missed it).

Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: There is one for no profiles yaml file. probably makes sense to add one for one with profiles too to ensure proper default selection. Adding now.

Comment (ivanporty) [reply skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/ui/SkaffoldProfilesComboBoxTest.kt]: done.
